public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataQueriesTab.tsx

```diff
@@ -358,9 +352,35 @@ export function PanelDataQueriesTabRendered({ model }: SceneComponentProps<Panel
             {queryLibraryEnabled && (
               <Button
                 icon="plus"
-                onClick={() => {
-                  openQueryLibraryDrawer(getDatasourceNames(datasource, queries), addQueryActionButton);
-                }}
+                onClick={() =>
+                  openQueryLibraryDrawer(getDatasourceNames(datasource, queries), async (query) => {
+                    // ensure all queries explicitly define a datasource
+                    const enrichedQueries = queries.map((q) =>
+                      q.datasource
+                        ? q
+                        : {
+                            ...q,
+                            datasource: datasource.getRef(),
+                          }
+                    );
+                    const newQueries = addQuery(enrichedQueries, query);
+                    model.onQueriesChange(newQueries);
+                    if (query.datasource?.uid) {
+                      const uniqueDatasources = new Set(newQueries.map((q) => q.datasource?.uid));
```

Comment:
Also maybe even moving this code into `model.onQueriesChange`? Can't imagine when changing to mixed if queries are mixed wouldn't be a desired behaviour when changing queries.

Reply:
hm, i'm a little hesitant to do that immediately as it will affect everything, and won't be feature toggled 😬 maybe in a follow up PR?